TauAnalysis/MCEmbeddingTools/plugins/CollectionMerger.cc

```diff
@@ -0,0 +1,215 @@
+#include "TauAnalysis/MCEmbeddingTools/plugins/CollectionMerger.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+
+#include "DataFormats/SiPixelCluster/interface/SiPixelCluster.h"
+#include "DataFormats/SiStripCluster/interface/SiStripCluster.h"
+
+#include "DataFormats/EcalRecHit/interface/EcalRecHit.h"
+#include "DataFormats/Common/interface/SortedCollection.h"
+
+
+#include "DataFormats/MuonDetId/interface/DTLayerId.h"
+#include "DataFormats/DTRecHit/interface/DTSLRecCluster.h"
+#include "DataFormats/DTRecHit/interface/DTRecHit1DPair.h"
+
+#include "DataFormats/MuonDetId/interface/CSCDetId.h"
+#include "DataFormats/CSCRecHit/interface/CSCRecHit2D.h"
+#include "DataFormats/MuonDetId/interface/RPCDetId.h"
+#include "DataFormats/RPCRecHit/interface/RPCRecHit.h"
+#include "DataFormats/HcalRecHit/interface/ZDCRecHit.h"
+
+
+#include "DataFormats/Common/interface/RangeMap.h"
+#include "DataFormats/Common/interface/OwnVector.h"
+
+typedef CollectionMerger<edmNew::DetSetVector<SiPixelCluster>, SiPixelCluster> PixelColMerger;
+typedef CollectionMerger<edmNew::DetSetVector<SiStripCluster>, SiStripCluster> StripColMerger;
+
+typedef CollectionMerger<edm::SortedCollection<EcalRecHit>, EcalRecHit> EcalRecHitColMerger;
+typedef CollectionMerger<edm::SortedCollection<HBHERecHit>, HBHERecHit> HBHERecHitColMerger;
+typedef CollectionMerger<edm::SortedCollection<HFRecHit>, HFRecHit> HFRecHitColMerger;
+typedef CollectionMerger<edm::SortedCollection<HORecHit>, HORecHit> HORecHitColMerger;
+typedef CollectionMerger<edm::SortedCollection<CastorRecHit>, CastorRecHit> CastorRecHitColMerger;
+typedef CollectionMerger<edm::SortedCollection<ZDCRecHit>, ZDCRecHit> ZDCRecHitColMerger;
+
+typedef CollectionMerger<edm::RangeMap<DTLayerId, edm::OwnVector<DTRecHit1DPair> >, DTRecHit1DPair> DTRecHitColMerger;
+typedef CollectionMerger<edm::RangeMap<CSCDetId, edm::OwnVector<CSCRecHit2D> >, CSCRecHit2D> CSCRecHitColMerger;
+typedef CollectionMerger<edm::RangeMap<RPCDetId, edm::OwnVector<RPCRecHit> >, RPCRecHit> RPCRecHitColMerger;
+
+
+
+// -------- Here Tracker Merger -----------
+template <typename T1, typename T2>
+void  CollectionMerger<T1,T2>::fill_output_obj_tracker(std::unique_ptr<MergeCollection > & output, std::vector<edm::Handle<MergeCollection> > &inputCollections, bool print_pixel)
+{
+
+    std::map<uint32_t, std::vector<BaseHit> >   output_map;
+  // First merge the collections with the help of the output map
+  for (auto inputCollection : inputCollections){
+   for ( typename MergeCollection::const_iterator clustSet = inputCollection->begin(); clustSet!= inputCollection->end(); ++clustSet ) {
+      DetId detIdObject( clustSet->detId() );
+      for (typename edmNew::DetSet<BaseHit>::const_iterator clustIt = clustSet->begin(); clustIt != clustSet->end(); ++clustIt ) { 	
+        output_map[detIdObject.rawId()].push_back(*clustIt);	
```

Comment:
Is there some reclustering code elsewhere?
Otherwise, how do you treat situations with overlapping clusters?